### PR TITLE
Making ENV Ractor-safe

### PR DESCRIPTION
**Pull request by rm155**
It is important that the ENV object be accessible in Ractors. To make it Ractor-safe, its methods that get/set its values are now synchronized with this modification. This is done by placing a mutex around each of these methods. With this, the ENV object is now Ractor-safe and shareable. Additionally, tests are added to ensure that its behavior is correct in Ractors.

### CODE REVIEW

**Comment by ioquatix** [comment]
Is there any performance impact for non-Ractor code?

**Comment by ioquatix** [comment]
Also, does this mean you can use `ENV` for message passing between Ractors!?

**Comment by rm155** [comment]
> 
> 
> Is there any performance impact for non-Ractor code?

There is some performance impact, but the level of impact varies between methods. For some methods, it seems negligible. For others, there is a visible impact on speed, but for many of these the impact is comparatively insignificant when ENV is larger (because the overhead from the locks remains constant).

**Comment by rm155** [comment]
> 
> 
> Also, does this mean you can use `ENV` for message passing between Ractors!?

Yes, it would be possible to have one Ractor store a value/message in ENV and have another Ractor read it.

File: hash.c [hunk 1]
```diff
@@ -4825,6 +4827,17 @@ extern char **environ;
 #define ENVNMATCH(s1, s2, n) (memcmp((s1), (s2), (n)) == 0)
 #endif
 
+rb_nativethread_lock_t env_lock;
+
+const char*
```
Comment (nobu) [inline hash.c:4832]:
```suggestion
static rb_nativethread_lock_t env_lock;

static const char*
```

File: hash.c [hunk 2]
```diff
@@ -5356,26 +5385,53 @@ env_aset(VALUE nm, VALUE val)
...
+{
+    char **env;
+    env = GET_ENVIRON(environ);
+    for (int i = 0; i < size; i++) {
+	const char *p = *env;
+	arr[i] = p;
+	env++;
+    }
+    FREE_ENVIRON(environ);
+}
```
Comment (nobu) [inline hash.c:5411]: Copied entries may no longer be valid once other threads change any envs.

Comment (nobu) [reply hash.c:5411]: More accurately, this function doesn't copy the env contents.

Comment (rm155) [reply hash.c:5411]: Thank you for your time in reviewing this!

I agree that the function is not creating a deep copy of the environment contents. However, in my tests, simply copying the pointer to the original environment was enough. know that inconsistencies in the data are possible, but I believe that ENV already has inconsistent behavior when used in simultaneous threads. Is there a case I missed that causes an error in the program?

Thanks again!

Comment (nobu) [reply hash.c:5411]: Therefore `ENV` isn't ractor-safe.
Before the ractor, GVL prevented it from the inconsistencies.

Comment (rm155) [reply hash.c:5411]: Thank you. I will think about it some more.

**Comment by ko1** [comment]
Discussion with @nobu we found that the shallow copy of environ is not enough for atomic operation. To avoid full-copy, we decided to create Ruby objects while locking. So I made new patch on this PR patches. #5263

@rm155 thank you for great contribution, I can't make new patch without your work.

**Pull request closed without merging**